// schema-org-graph-js@0.5.1/dist/runtime/base.mjs
import {
  addressResolver,
  aggregateOfferResolver,
  aggregateRatingResolver,
  articleResolver,
  bookEditionResolver,
  bookResolver,
  breadcrumbResolver,
  commentResolver,
  courseResolver,
  eventResolver,
  howToResolver,
  howToStepResolver,
  imageResolver,
  itemListResolver,
  localBusinessResolver,
  movieResolver,
  offerResolver,
  organizationResolver,
  personResolver,
  placeResolver,
  productResolver,
  provideResolver,
  questionResolver,
  readActionResolver,
  recipeResolver,
  resolveOpeningHours,
  reviewResolver,
  searchActionResolver,
  softwareAppResolver,
  videoResolver,
  virtualLocationResolver,
  webPageResolver,
  webSiteResolver
} from "schema-org-graph-js";
export const defineAddress = (input) => provideResolver(input, addressResolver);
export const defineAggregateOffer = (input) => provideResolver(input, aggregateOfferResolver);
export const defineAggregateRating = (input) => provideResolver(input, aggregateRatingResolver);
export const defineArticle = (input) => provideResolver(input, articleResolver);
export const defineBook = (input) => provideResolver(input, bookResolver);
export const defineBookEdition = (input) => provideResolver(input, bookEditionResolver);
export const defineBreadcrumb = (input) => provideResolver(input, breadcrumbResolver);
export const defineComment = (input) => provideResolver(input, commentResolver);
export const defineCourse = (input) => provideResolver(input, courseResolver);
export const defineEvent = (input) => provideResolver(input, eventResolver);
export const defineVirtualLocation = (input) => provideResolver(input, virtualLocationResolver);
export const definePlace = (input) => provideResolver(input, placeResolver);
export const defineHowTo = (input) => provideResolver(input, howToResolver);
export const defineHowToStep = (input) => provideResolver(input, howToStepResolver);
export const defineItemList = (input) => provideResolver(input, itemListResolver);
export const defineImage = (input) => provideResolver(input, imageResolver);
export const defineLocalBusiness = (input) => provideResolver(input, localBusinessResolver);
export const defineMovie = (input) => provideResolver(input, movieResolver);
export const defineOffer = (input) => provideResolver(input, offerResolver);
export const defineOpeningHours = (input) => provideResolver(input, resolveOpeningHours);
export const defineOrganization = (input) => provideResolver(input, organizationResolver);
export const definePerson = (input) => provideResolver(input, personResolver);
export const defineProduct = (input) => provideResolver(input, productResolver);
export const defineQuestion = (input) => provideResolver(input, questionResolver);
export const defineRecipe = (input) => provideResolver(input, recipeResolver);
export const defineSoftwareApp = (input) => provideResolver(input, softwareAppResolver);
export const defineReview = (input) => provideResolver(input, reviewResolver);
export const defineVideo = (input) => provideResolver(input, videoResolver);
export const defineWebPage = (input) => provideResolver(input, webPageResolver);
export const defineWebSite = (input) => provideResolver(input, webSiteResolver);
export const defineSearchAction = (input) => provideResolver(input, searchActionResolver);
export const defineReadAction = (input) => provideResolver(input, readActionResolver);
