// schema-org-graph-js@0.5.1/dist/runtime/base.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defineWebSite = exports.defineWebPage = exports.defineVirtualLocation = exports.defineVideo = exports.defineSoftwareApp = exports.defineSearchAction = exports.defineReview = exports.defineRecipe = exports.defineReadAction = exports.defineQuestion = exports.defineProduct = exports.definePlace = exports.definePerson = exports.defineOrganization = exports.defineOpeningHours = exports.defineOffer = exports.defineMovie = exports.defineLocalBusiness = exports.defineItemList = exports.defineImage = exports.defineHowToStep = exports.defineHowTo = exports.defineEvent = exports.defineCourse = exports.defineComment = exports.defineBreadcrumb = exports.defineBookEdition = exports.defineBook = exports.defineArticle = exports.defineAggregateRating = exports.defineAggregateOffer = exports.defineAddress = void 0;

var _schemaOrgGraphJs = require("schema-org-graph-js");

const defineAddress = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.addressResolver);

exports.defineAddress = defineAddress;

const defineAggregateOffer = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.aggregateOfferResolver);

exports.defineAggregateOffer = defineAggregateOffer;

const defineAggregateRating = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.aggregateRatingResolver);

exports.defineAggregateRating = defineAggregateRating;

const defineArticle = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.articleResolver);

exports.defineArticle = defineArticle;

const defineBook = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.bookResolver);

exports.defineBook = defineBook;

const defineBookEdition = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.bookEditionResolver);

exports.defineBookEdition = defineBookEdition;

const defineBreadcrumb = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.breadcrumbResolver);

exports.defineBreadcrumb = defineBreadcrumb;

const defineComment = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.commentResolver);

exports.defineComment = defineComment;

const defineCourse = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.courseResolver);

exports.defineCourse = defineCourse;

const defineEvent = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.eventResolver);

exports.defineEvent = defineEvent;

const defineVirtualLocation = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.virtualLocationResolver);

exports.defineVirtualLocation = defineVirtualLocation;

const definePlace = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.placeResolver);

exports.definePlace = definePlace;

const defineHowTo = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.howToResolver);

exports.defineHowTo = defineHowTo;

const defineHowToStep = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.howToStepResolver);

exports.defineHowToStep = defineHowToStep;

const defineItemList = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.itemListResolver);

exports.defineItemList = defineItemList;

const defineImage = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.imageResolver);

exports.defineImage = defineImage;

const defineLocalBusiness = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.localBusinessResolver);

exports.defineLocalBusiness = defineLocalBusiness;

const defineMovie = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.movieResolver);

exports.defineMovie = defineMovie;

const defineOffer = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.offerResolver);

exports.defineOffer = defineOffer;

const defineOpeningHours = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.resolveOpeningHours);

exports.defineOpeningHours = defineOpeningHours;

const defineOrganization = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.organizationResolver);

exports.defineOrganization = defineOrganization;

const definePerson = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.personResolver);

exports.definePerson = definePerson;

const defineProduct = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.productResolver);

exports.defineProduct = defineProduct;

const defineQuestion = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.questionResolver);

exports.defineQuestion = defineQuestion;

const defineRecipe = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.recipeResolver);

exports.defineRecipe = defineRecipe;

const defineSoftwareApp = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.softwareAppResolver);

exports.defineSoftwareApp = defineSoftwareApp;

const defineReview = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.reviewResolver);

exports.defineReview = defineReview;

const defineVideo = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.videoResolver);

exports.defineVideo = defineVideo;

const defineWebPage = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.webPageResolver);

exports.defineWebPage = defineWebPage;

const defineWebSite = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.webSiteResolver);

exports.defineWebSite = defineWebSite;

const defineSearchAction = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.searchActionResolver);

exports.defineSearchAction = defineSearchAction;

const defineReadAction = input => (0, _schemaOrgGraphJs.provideResolver)(input, _schemaOrgGraphJs.readActionResolver);

exports.defineReadAction = defineReadAction;
