// @vueuse/schema-org/dist/index.mjs
export { A as AliasRuntime, P as PkgName, S as SchemaOrgResolver, r as resolveUserConfig, s as schemaAutoImports, a as schemaOrgAutoImports, b as schemaOrgComponents } from './shared/schema-org.714403e2.mjs';
import { isRef, unref, ref, computed, inject } from 'vue';
import { createSchemaOrgGraph, resolveMeta, buildResolvedGraphCtx, organiseNodes, renderNodesToSchemaOrgHtml } from 'schema-org-graph-js';

const isObject = (val) => val !== null && typeof val === "object";
const isArray = Array.isArray;
const smartUnref = (val) => {
  if (val !== null && !isRef(val) && typeof val === "object")
    return deepUnref(val);
  return unref(val);
};
const unrefArray = (arr) => {
  const unreffed = [];
  arr.forEach((val) => {
    unreffed.push(smartUnref(val));
  });
  return unreffed;
};
const unrefObject = (obj) => {
  const unreffed = {};
  Object.keys(obj).forEach((key) => {
    unreffed[key] = smartUnref(obj[key]);
  });
  return unreffed;
};
const deepUnref = (val) => {
  const checkedVal = isRef(val) ? unref(val) : val;
  if (!isObject(checkedVal))
    return checkedVal;
  if (isArray(checkedVal))
    return unrefArray(checkedVal);
  return unrefObject(checkedVal);
};

const createSchemaOrg = (options) => {
  if (!options.position)
    options.position = "body";
  const schemaRef = ref("");
  let ctx = createSchemaOrgGraph();
  const resolveGraphNodesToHtml = async () => {
    const meta = await options.meta();
    const resolvedMeta = resolveMeta(deepUnref(meta));
    const resolvedCtx = buildResolvedGraphCtx(ctx.nodes.map(deepUnref), resolvedMeta);
    const nodes = organiseNodes(resolvedCtx.nodes);
    return renderNodesToSchemaOrgHtml(nodes);
  };
  const client = {
    ctx,
    options,
    install(app) {
      app.config.globalProperties.$schemaOrg = client;
      app.provide("schemaorg", client);
    },
    async generateSchema() {
      schemaRef.value = await resolveGraphNodesToHtml();
      return schemaRef;
    },
    async forceRefresh() {
      await client.generateSchema();
      await client.setupDOM();
    },
    removeContext(uid) {
      const newCtx = createSchemaOrgGraph();
      newCtx.meta = ctx.meta;
      newCtx.addNode(ctx.nodes.filter((n) => n._uid !== uid));
      ctx = newCtx;
    },
    setupDOM() {
      return options.updateHead(computed(() => {
        const script = {
          "type": "application/ld+json",
          "data-id": "schema-org-graph",
          "key": "schema-org-graph",
          "children": schemaRef.value
        };
        if (options.position === "body")
          script.body = true;
        return {
          script: [
            script
          ]
        };
      }));
    }
  };
  return client;
};

function injectSchemaOrg() {
  let client;
  try {
    client = inject("schemaorg");
  } catch (e) {
  }
  if (!client)
    console.warn("[@vueuse/schema-org] Failed to find plugin, you may have forgotten to apply app.use(schemaOrg)");
  return client;
}

export { createSchemaOrg, injectSchemaOrg };
